### PR TITLE
Coin-tester tool engine + EVM coin-tester

### PR DESCRIPTION
**Pull request by lvndry**
### ✅ Checklist

- [x] `npx changeset` was attached.
- [x] **Covered by automatic tests.** 
- [x] **Impact of the changes:** 
  - ...

### 📝 Description

This PR is the first version of the non regression tool used to test our ledger's coin-module. The tool runs a series of scenarios similar to how a user can interact with the modules on the different clients.
The tool can be run locally and on the CI easily.

### ❓ Context

- **JIRA or GitHub link**: 

---

### 🧐 Checklist for the PR Reviewers

- **The code aligns with the requirements** described in the linked JIRA or GitHub issue.
- **The PR description clearly documents the changes** made and explains any technical trade-offs or design decisions.
- **There are no undocumented trade-offs**, technical debt, or maintainability issues.
- **The PR has been tested** thoroughly, and any potential edge cases have been considered and handled.
- **Any new dependencies** have been justified and documented.
- **Performance** considerations have been taken into account. (changes have been profiled or benchmarked if necessary)

### CODE REVIEW

File: libs/coin-modules/coin-evm/src/__tests__/coin-tester/docker.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+      ...process.env,
+      RPC: rpc,
+    },
+  });
+
+  const checkAnvilLogs = async (): Promise<void> => {
+    const { out } = await compose.logs("anvil", { cwd });
+
+    if (out.includes("Listening on 0.0.0.0:")) {
+      console.log(" -  ANVIL READY ✅  - ");
```
Comment (lambertkevin) [inline libs/coin-modules/coin-evm/src/__tests__/coin-tester/docker.ts:23]: If you don't mind, I'd like to keep some "Chalk" here and there in the logs for important informations

File: libs/coin-modules/coin-evm/src/__tests__/coin-tester/indexer.ts
Comment (lambertkevin) [inline libs/coin-modules/coin-evm/src/__tests__/coin-tester/indexer.ts:1]: Just for information, we'll need to work on adding 1155 events as well + doing the same thing with Etherscan like explorers, otherwise we won't be able to run the tool on anything but Ethereum Polygon & BSC

File: libs/coin-modules/coin-evm/src/__tests__/coin-tester/scenario-1.test.ts [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+    await provider.waitForTransaction(boredApeTxHash);
+
+    await getLogs();
+
+    return { currencyBridge, accountBridge, account: scenarioAccount, onSignerConfirmation };
+  },
+  afterEach: () => {
+    clearExplorerAppendix();
+  },
+  transactions: [scenarioSendTransaction, scenarioERC721Transaction],
```
Comment (lambertkevin) [inline libs/coin-modules/coin-evm/src/__tests__/coin-tester/scenario-1.test.ts:174]: Missing 1 or more ERC20 transactions

File: libs/coin-modules/coin-evm/src/prepareTransaction.ts [hunk 3]
```diff
@@ -59,6 +59,7 @@ const prepareCoinTransaction = async (
     // and displayed in the UI as `set the gas manually`
     () => new BigNumber(0),
   );
+
```
Comment (lambertkevin) [inline libs/coin-modules/coin-evm/src/prepareTransaction.ts:62]: Unnecessary change

File: libs/coin-tester/.env.example
Comment (lambertkevin) [inline libs/coin-tester/.env.example:1]: Like mentioned before, would be very cool to look for an option to only have 1 .env file, located in the non-reg folder of a code module

File: libs/coin-tester/docker-compose.yml [hunk 4]
```diff
@@ -0,0 +1,18 @@
+version: '2.4'
+
+services:
+  speculos:
+    # Speculos isn't supported for Mac M1 yet. You have to build the image locally as described in the README.
+    # Otherwise you can use the official image
+    image: ghcr.io/ledgerhq/speculos:latest
```
Comment (lambertkevin) [inline libs/coin-tester/docker-compose.yml:7]: We should look to have an env variable here to chose the right imagine depending on M1 or CI

File: libs/coin-tester/src/docker.ts [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+import SpeculosTransportHttp from "@ledgerhq/hw-transport-node-speculos-http";
+import { ENV } from "./types";
+
+const { API_PORT } = process.env as ENV;
+const cwd = path.join(__dirname);
+
+const delay = (timing: number) => new Promise(resolve => setTimeout(resolve, timing));
+
+// type SemanticVersion = `${number}.${number}.${number}`;
+// type NanoAppEndpoint = `/${SemanticVersion}/${string}/app_${SemanticVersion}.elf`;
```
Comment (lambertkevin) [inline libs/coin-tester/src/docker.ts:14]: Still necessary ?

File: libs/coin-tester/src/docker.ts [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+    env: {
+      ...process.env,
+    },
+  });
+
+  const checkSpeculosLogs = async (): Promise<SpeculosTransportHttp> => {
+    const { out } = await compose.logs("speculos", { cwd });
+
+    if (out.includes("Running on all addresses (0.0.0.0)")) {
+      console.log(" -  SPECULOS READY ✅  - ");
```
Comment (lambertkevin) [inline libs/coin-tester/src/docker.ts:47]: Same here with Chalk 🙏

File: libs/coin-modules/coin-evm/src/__tests__/coin-tester/scenario-1.test.ts [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+            action: "press-and-release",
+          });
+          onSignerConfirmation(e);
+        } else {
+          await axios.post(`http://localhost:${process.env.API_PORT}/button/both`, {
+            action: "press-and-release",
+          });
+        }
+      }
+    };
```
Comment (lambertkevin) [inline libs/coin-modules/coin-evm/src/__tests__/coin-tester/scenario-1.test.ts:84]: Still think we should extract that from the setup, it feels messy here I think 🤔

File: libs/ledgerjs/packages/hw-app-eth/src/services/ledger/nfts.ts [hunk 8]
```diff
@@ -28,6 +28,7 @@ export const getNFTInfo = async (
       log("error", "could not fetch from " + url + ": " + String(e));
       return null;
     });
+
```
Comment (lambertkevin) [inline libs/ledgerjs/packages/hw-app-eth/src/services/ledger/nfts.ts:31]: Unnecessary change

File: libs/ledgerjs/packages/hw-app-eth/src/services/ledger/nfts.ts [hunk 9]
```diff
@@ -63,6 +64,7 @@ export const loadNftPlugin = async (
...
```
Comment (lambertkevin) [inline libs/ledgerjs/packages/hw-app-eth/src/services/ledger/nfts.ts:67]: Unnecessary change

**Review by lambertkevin — CHANGES_REQUESTED** [review]
Good job, a bunch of stuffs are not ready to merge yet, and we can discuss about it, but it's a good first step.

Also please, squash all your commits and undo your last commit to recreate clean commits of everything (and remove my "temp" ones as well 😛)

File: libs/coin-modules/coin-evm/src/.env.example
[the remark this answers is not in the document]
Comment (lvndry) [reply libs/coin-modules/coin-evm/src/.env.example:1]: The issue is the .env file needs to be located with the `jest.config.js` where we call the 'dotenv/config'

**Review by lambertkevin — COMMENTED** [review]
New temporary comments so you can fix it while doing the squash thingy 😋

File: libs/coin-modules/coin-evm/.eslintrc.js
[the remark this answers is not in the document]
Comment (lvndry) [reply libs/coin-modules/coin-evm/.eslintrc.js:1]: The linter didn't change per say, I only moved rules up in the files and said that `console.xxx` is now allowed in test files

File: libs/coin-modules/coin-evm/src/__tests__/coin-tester/docker.ts [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+const cwd = __dirname;
+
+const delay = (timing: number): Promise<void> =>
+  new Promise(resolve => setTimeout(resolve, timing));
+
+export const spawnAnvil = async (rpc: string): Promise<void> => {
+  console.log("Starting anvil...");
+  await compose.upOne("anvil", {
+    cwd,
+    // log: true,
```
Comment (lambertkevin) [inline libs/coin-modules/coin-evm/src/__tests__/coin-tester/docker.ts:14]: Reactive the logs please 🙏

File: libs/coin-modules/coin-evm/src/__tests__/coin-tester/docker.ts [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+  };
+
+  await checkAnvilLogs();
+};
+
+export const killAnvil = async (): Promise<void> => {
+  console.log("Stopping anvil...");
+  await compose.down({
+    cwd,
+    // log: true,
```
Comment (lambertkevin) [inline libs/coin-modules/coin-evm/src/__tests__/coin-tester/docker.ts:46]: Reactive the logs please 🙏

File: libs/coin-tester/src/docker.ts [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+      Authorization: `Bearer ${process.env.GH_TOKEN}`,
+    },
+  });
+
+  await fs.mkdir(path.resolve(cwd, "tmp"), { recursive: true });
+  await fs.writeFile(path.resolve(cwd, "tmp/app.elf"), blob, "binary");
+
+  await compose.upOne("speculos", {
+    cwd,
+    // log: true,
```
Comment (lambertkevin) [inline libs/coin-tester/src/docker.ts:38]: Reactive the logs please 🙏

File: libs/coin-tester/src/main.ts [hunk 13]
```diff
@@ -0,0 +1,184 @@
...
+    accountBridge: AccountBridge<T>;
+    currencyBridge: CurrencyBridge;
+    account: Account;
+    testTimeout?: number;
+    retryInterval?: number;
+    onSignerConfirmation?: (e?: SignOperationEvent) => Awaited<void>;
+  }>;
+  transactions: ScenarioTransaction<Partial<T>>[];
+  beforeAll?: () => Awaited<void>;
+  afterAll?: () => Awaited<void>;
```
Comment (lambertkevin) [inline libs/coin-tester/src/main.ts:29]: I think we kinda missed that we're not testing the account before and after the scenario execution. 
`beforeAll` & `afterAll` should have `account` as parameter and the scenario should be able to set some `expect` here based on it

File: libs/coin-modules/coin-evm/src/__tests__/coin-tester/scenarios/ethereum.ts [hunk 14]
```diff
@@ -0,0 +1,200 @@
...
+  ERC1155Interface,
+} from "../helpers";
+import { clearExplorerAppendix, getLogs, setBlock } from "../indexer";
+import { killAnvil, spawnAnvil } from "../docker";
+
+const scenarioSendTransaction: ScenarioTransaction<Partial<EvmTransaction>> = {
+  name: "Send ethereum",
+  amount: new BigNumber(100),
+  recipient: "0x6bfD74C0996F269Bcece59191EFf667b3dFD73b9",
+};
```
Comment (lambertkevin) [inline libs/coin-modules/coin-evm/src/__tests__/coin-tester/scenarios/ethereum.ts:27]: We forgot to add expect to each transaction in order to verify the result of the sync after each broadcast.

File: libs/coin-modules/coin-evm/src/__tests__/coin-tester/scenarios/ethereum.ts [hunk 15]
```diff
@@ -0,0 +1,200 @@
...
+// use function createTransaction
+const scenarioUSDCTransaction: ScenarioTransaction<Partial<EvmTransaction>> = {
+  name: "Send USDC",
+  amount: new BigNumber(100),
+  recipient: "0x6bfD74C0996F269Bcece59191EFf667b3dFD73b9",
+  subAccountId: encodeTokenAccountId(
+    "js:2:ethereum:0x3313797c7B45F34c56Bdedc0179992A4d435AF25:",
+    USDC_ON_ETHEREUM,
+  ),
+};
```
Comment (lambertkevin) [inline libs/coin-modules/coin-evm/src/__tests__/coin-tester/scenarios/ethereum.ts:38]: same thing here

File: libs/coin-modules/coin-evm/src/__tests__/coin-tester/scenarios/ethereum.ts [hunk 16]
```diff
@@ -0,0 +1,200 @@
...
+    name: "Send NFT",
+    recipient: "0x6bfD74C0996F269Bcece59191EFf667b3dFD73b9",
+    mode: "erc721",
+    nft: {
+      tokenId: "3368",
+      contract: "0xBC4CA0EdA7647A8aB7C2061c2E118A18a936f13D",
+      quantity: new BigNumber(1),
+      collectionName: "Bored Ape",
+    },
+  };
```
Comment (lambertkevin) [inline libs/coin-modules/coin-evm/src/__tests__/coin-tester/scenarios/ethereum.ts:51]: same thing here

File: libs/coin-modules/coin-evm/src/__tests__/coin-tester/scenarios/ethereum.ts [hunk 17]
```diff
@@ -0,0 +1,200 @@
...
+    recipient: "0x6bfD74C0996F269Bcece59191EFf667b3dFD73b9",
+    mode: "erc1155",
+    nft: {
+      tokenId: "951",
+      collectionName: "Clone X",
+      // EIP55 checksum of 0x348fc118bcc65a92dc033a951af153d14d945312
+      contract: "0x348FC118bcC65a92dC033A951aF153d14D945312",
+      quantity: new BigNumber(2),
+    },
+  };
```
Comment (lambertkevin) [inline libs/coin-modules/coin-evm/src/__tests__/coin-tester/scenarios/ethereum.ts:65]: same thing here

File: libs/coin-modules/coin-evm/src/__tests__/coin-tester/scenarios/ethereum.ts [hunk 18]
```diff
@@ -0,0 +1,200 @@
...
+
+const defaultNanoApp = { firmware: "2.2.3" as const, version: "1.10.4" as const };
+
+export const scenarioEthereum: Scenario<EvmTransaction> = {
```
Comment (lambertkevin) [inline libs/coin-modules/coin-evm/src/__tests__/coin-tester/scenarios/ethereum.ts:69]: Scenario should be implementing a `beforeAll` and `afterAll` acknowledging the account state ("account has X balance, Y tokens, Z NFTs"...) to confirm the correct execution of the smart contracts

File: libs/coin-modules/coin-evm/src/__tests__/coin-tester/scenarios/polygon.ts
Comment (lambertkevin) [inline libs/coin-modules/coin-evm/src/__tests__/coin-tester/scenarios/polygon.ts:1]: And same thing to apply here + missing ERC721 + ERC1155 transactions.

**Review by lambertkevin — COMMENTED** [review]
Still missing some things, but should be ready after that 🚀

File: libs/coin-tester/src/docker.ts [hunk 19]
```diff
@@ -0,0 +1,96 @@
...
+export const spawnSigner = async (
+  service: "speculos",
+  nanoAppEndpoint: `/${string}`,
+): Promise<{
+  transport: SpeculosTransportHttp;
+  onSignerConfirmation: (e?: SignOperationEvent) => Promise<void>;
+}> => {
+  console.log(`Starting ${service}...`);
+
+  const { data: blob } = await axios({
```
Comment (sprohaszka-ledger) [inline libs/coin-tester/src/docker.ts:24]: I would use `live-network` instead of directly `axios`

File: libs/coin-tester/src/main.ts [hunk 20]
```diff
@@ -0,0 +1,186 @@
...
+    const optimisticOperation = await accountBridge.broadcast({
+      account: scenarioAccount,
+      signedOperation,
+    });
+
+    console.log(" → ", "🛫 ", chalk.bold("Broadcated the transaction"), "✓");
+
+    const rety_limit = 10;
+
+    const expcectHandler = async (retry = rety_limit) => {
```
Comment (sprohaszka-ledger) [inline libs/coin-tester/src/main.ts:137]: **typo** `expcectHandler` --> `expectHandler`

File: package.json [hunk 21]
```diff
@@ -62,6 +62,7 @@
     "coin:bitcoin": "pnpm --filter coin-bitcoin",
     "coin:evm": "pnpm --filter coin-evm",
     "coin:framework": "pnpm --filter coin-framework",
+    "coin:tester": "pnpm --filter coin-tester",
```
Comment (sprohaszka-ledger) [inline package.json:65]: Is it really necessary as it seems not to work without coin-module implement it?

File: libs/coin-tester/package.json [hunk 22]
```diff
@@ -0,0 +1,78 @@
...
+    "build": "tsc && tsc -m ES6 --outDir lib-es",
+    "prewatch": "pnpm build",
+    "watch": "tsc --watch",
+    "doc": "documentation readme src/** --section=API --pe ts --re ts --re d.ts",
+    "lint": "eslint ./src --no-error-on-unmatched-pattern --ext .ts,.tsx --cache",
+    "lint:fix": "pnpm lint --fix",
+    "test": "jest"
+  },
+  "dependencies": {
+    "@ledgerhq/coin-framework": "workspace:^",
```
Comment (sprohaszka-ledger) [inline libs/coin-tester/package.json:17]: Those dependencies are not necessary: `coin-framework`, `cryptoassets`, `hw-app-eth`, `hw-transport`.
Could you check what is the minimum required (I see also bignumber and rxjs and not sure where they are used).

File: libs/coin-modules/coin-evm/src/__tests__/coin-tester/scenarios/ethereum.ts [hunk 23]
```diff
@@ -0,0 +1,204 @@
...
+      currency: getCryptoCurrencyById("ethereum"),
+      derivationMode: "",
+    });
+
+    const scenarioAccount = makeAccount(address, ethereum);
+
+    await setBlock();
+
+    // USDC
+    const addressToImpersonateBinance = "0x47ac0Fb4F2D84898e4D9E7b4DaB3C24507a6D503"; // Binance
```
Comment (sprohaszka-ledger) [inline libs/coin-modules/coin-evm/src/__tests__/coin-tester/scenarios/ethereum.ts:118]: It could help readability to extract USDC, BoredApe and CloneX into functions.

File: libs/coin-modules/coin-evm/src/__tests__/coin-tester/indexer.ts [hunk 24]
```diff
@@ -0,0 +1,220 @@
...
+  const toBlock = await provider.getBlockNumber();
+  const logs = await provider.getLogs({
+    fromBlock,
+    toBlock,
+    topics: [
+      [TRANSFER_EVENTS_TOPICS.ERC20, TRANSFER_EVENTS_TOPICS.ERC721, TRANSFER_EVENTS_TOPICS.ERC1155],
+    ],
+  });
+
+  if (logs.length) {
```
Comment (sprohaszka-ledger) [inline libs/coin-modules/coin-evm/src/__tests__/coin-tester/indexer.ts:46]: Unnecessary check. If there is no `logs`, the loop will do nothing

File: libs/coin-modules/coin-evm/src/__tests__/coin-tester/indexer.ts [hunk 25]
```diff
@@ -0,0 +1,220 @@
...
+        explorerAppendixByAddress[to] = new Map();
+      }
+      explorerAppendixByAddress[to].set(operation.hash, operation);
+    }
+  }
+
+  return logs;
+};
+
+export const indexBlocks = async (): Promise<void> => {
```
Comment (sprohaszka-ledger) [inline libs/coin-modules/coin-evm/src/__tests__/coin-tester/indexer.ts:149]: I don't see where this function is used 🤔

**Review by sprohaszka-ledger — CHANGES_REQUESTED** [review]
See if you can remove unused dependencies.

File: package.json (hunk 21 above)
Comment (lvndry) [reply package.json:65]: It's a helper for development so that we can run scripts inside the module like `pnpm coin:tester build` `pnpm coin:tester watch` directly from the root

File: libs/coin-tester/src/docker.ts (hunk 19 above)
Comment (lvndry) [reply libs/coin-tester/src/docker.ts:24]: Can be discussed but I'd rather not as it was though for Ledger Live

File: libs/coin-modules/coin-evm/src/__tests__/coin-tester/indexer.ts (hunk 25 above)
Comment (lvndry) [reply libs/coin-modules/coin-evm/src/__tests__/coin-tester/indexer.ts:147]: Yes not used *yet* but let's keep it as it'll be useful soon

File: apps/ledger-live-desktop/src/renderer/screens/market/index.tsx [hunk 26]
```diff
@@ -89,7 +89,7 @@ export default function Market() {
   useEffect(() => {
     const intervalId = setInterval(() => refetchData(marketCurrentPage ?? 1), refreshRate);
 
-    return () => clearInterval(intervalId);
+    return () => clearInterval(intervalId as unknown as number);
```
Comment (lvndry) [inline apps/ledger-live-desktop/src/renderer/screens/market/index.tsx:92]: Not related to the PR but the CI failed because of this: https://github.com/LedgerHQ/ledger-live/actions/runs/8805039576/job/24166776064#step:7:28

File: libs/coin-modules/coin-evm/src/__tests__/coin-tester/docker-compose.yml [hunk 27]
```diff
@@ -0,0 +1,9 @@
+version: '2.4'
```
[the remark this answers is not in the document]
Comment (lambertkevin) [reply libs/coin-modules/coin-evm/src/__tests__/coin-tester/docker-compose.yml]: Actually docker-compose v2 and v3 are for different purposes, v3 is thought to be used with Docker Swarm in mind, while v2 is still the preferred solution for dev not requiring orchestration as some specific features from v3 won't actually work with local or basic deployment of the containers.

File: libs/coin-modules/coin-evm/src/__tests__/coin-tester/anvil.ts [hunk 28]
```diff
@@ -0,0 +1,55 @@
+import { v2 as compose } from "docker-compose";
```
Comment (lambertkevin) [inline libs/coin-modules/coin-evm/src/__tests__/coin-tester/anvil.ts:1]: If we still want to upgrade to docker-compose v3, that should probably be changed as well.

File: libs/coin-modules/coin-evm/src/__tests__/coin-tester/docker-compose.yml [hunk 29]
```diff
@@ -0,0 +1,9 @@
+version: '3.8'
```
Comment (lambertkevin) [inline libs/coin-modules/coin-evm/src/__tests__/coin-tester/docker-compose.yml:1]: To be checked if we really want to upgrade to v3 as, as I mentionned on Stephane's comment, [v2 is still the recommended version for development without docker swarm](https://forums.docker.com/t/limit-resources-without-swarm/98524/2).

File: libs/coin-modules/coin-evm/src/__tests__/coin-tester/scenarios/ethereum.ts [hunk 30]
```diff
@@ -0,0 +1,252 @@
...
+
+const scenarioSendTransaction: ScenarioTransaction<EvmTransaction> = {
+  name: "Send ethereum",
+  amount: new BigNumber(100),
+  recipient: "0x6bfD74C0996F269Bcece59191EFf667b3dFD73b9",
+  expect: account => {
+    const { operations } = account;
+    const sendOperation = operations.find(operation => operation.transactionSequenceNumber === 0);
+    expect(sendOperation?.transactionRaw?.amount === "100").toBeDefined();
+  },
```
Comment (lambertkevin) [inline libs/coin-modules/coin-evm/src/__tests__/coin-tester/scenarios/ethereum.ts:31]: Not sure this test is really what we want to verify. I think we can check 2 things:
- Account should have 1 more operation compared to our initial situation (I guess account has 3 operations at the beginning, so account.operations.length should now be 4 ?).
- Latest operation has type OUT, value is 100 + operation.fee
- Balance is now 10000^18 (because we took an account initialized by anvil with 10k ETH balance) - 100 (operation value) - operation.fee (will be a little bit different when we're testing L2s, but that's a problem for tomorrow :p)

File: libs/coin-modules/coin-evm/src/__tests__/coin-tester/scenarios/ethereum.ts [hunk 31]
```diff
@@ -0,0 +1,252 @@
...
+  recipient: "0x6bfD74C0996F269Bcece59191EFf667b3dFD73b9",
+  subAccountId: encodeTokenAccountId(
+    "js:2:ethereum:0x3313797c7B45F34c56Bdedc0179992A4d435AF25",
+    USDC_ON_ETHEREUM,
+  ),
+  expect: account => {
+    const { operations } = account;
+    const sendOperation = operations.find(operation => operation.transactionSequenceNumber === 1);
+    expect(sendOperation?.transactionRaw?.amount).toBe("80");
+  },
```
Comment (lambertkevin) [inline libs/coin-modules/coin-evm/src/__tests__/coin-tester/scenarios/ethereum.ts:47]: Same here but even easier:
- TokenAccount balance is now 100 - 80
- Account has now 6 ops (if my initial assumption of account having 3 ops after setup)
- Latest op should have type FEES & 1 subOperation of type OUT and value of 80^6 & contract is USDC's address

File: libs/coin-modules/coin-evm/src/__tests__/coin-tester/scenarios/ethereum.ts [hunk 32]
```diff
@@ -0,0 +1,252 @@
...
+    const sendOperation = account.operations.find(
+      operation => operation.transactionSequenceNumber === 2,
+    );
+
+    expect(
+      sendOperation?.nftOperations?.find(
+        op => op.contract === "0xBC4CA0EdA7647A8aB7C2061c2E118A18a936f13D",
+      ),
+    ).toBeDefined();
+  },
```
Comment (lambertkevin) [inline libs/coin-modules/coin-evm/src/__tests__/coin-tester/scenarios/ethereum.ts:70]: Same here:

- Account.nfts should not contain an NFT for `0xBC4CA0EdA7647A8aB7C2061c2E118A18a936f13D` anymore
- Account has now 7 ops (if my initial assumption of account having 3 ops after setup)
- Latest op should have type FEES & 1 nftOperation of type NFT_OUT, nftOperation.contract should be `0xBC4CA0EdA7647A8aB7C2061c2E118A18a936f13D` and value I guess is 1 for ERC721 transfers.

File: libs/coin-modules/coin-evm/src/__tests__/coin-tester/scenarios/ethereum.ts [hunk 33]
```diff
@@ -0,0 +1,252 @@
...
+  expect: account => {
+    const sendOperation = account.operations.find(
+      operation => operation.transactionSequenceNumber === 3,
+    );
+    expect(
+      sendOperation?.nftOperations?.find(
+        op => op.contract === "0x348FC118bcC65a92dC033A951aF153d14D945312",
+      ),
+    ).toBeDefined();
+  },
```
Comment (lambertkevin) [inline libs/coin-modules/coin-evm/src/__tests__/coin-tester/scenarios/ethereum.ts:92]: Same here:

- Account.nfts should not contain an NFT for `0x348FC118bcC65a92dC033A951aF153d14D945312` anymore
- Account has now 8 ops (if my initial assumption of account having 3 ops after setup)
- Latest op should have type FEES & 1 nftOperation of type NFT_OUT, nftOperation.contract should be `0x348FC118bcC65a92dC033A951aF153d14D945312` and value is 2

File: libs/coin-modules/coin-evm/src/__tests__/coin-tester/scenarios/ethereum.ts [hunk 34]
```diff
@@ -0,0 +1,252 @@
...
+    type: "0x0",
+    gasPrice: (await provider.getGasPrice()).toHexString(),
+    nonce: "0x" + (await provider.getTransactionCount(addressToImpersonateBinance)).toString(16),
+    chainId: "0x" + (await provider.getNetwork()).chainId.toString(16),
+  };
+
+  const hash = await provider.send("eth_sendTransaction", [sendUSDC]);
+  await provider.send("anvil_stopImpersonatingAccount", [addressToImpersonateBinance]);
+  await provider.waitForTransaction(hash);
+};
```
Comment (lambertkevin) [inline libs/coin-modules/coin-evm/src/__tests__/coin-tester/scenarios/ethereum.ts:117]: Just a comment for later, but rather than having those unique methods for each init, I guess we can create a helper to "impersonnate" an account and just execute a random ethers.Transaction in its name, because we're doing the same thing every time here 🤷‍♂️

File: libs/coin-modules/coin-evm/src/__tests__/coin-tester/scenarios/ethereum.ts [hunk 35]
```diff
@@ -0,0 +1,252 @@
...
+    expect(account.operations.filter(operation => operation.type === "OUT").length).toBe(2);
+    expect(
+      account.operations
+        .filter(operation => operation.type === "FEES")
+        .filter(
+          feesOperations =>
+            feesOperations.nftOperations?.find(nftOperation => nftOperation.type === "NFT_OUT"),
+        ).length,
+    ).toBe(2);
+  },
```
Comment (lambertkevin) [inline libs/coin-modules/coin-evm/src/__tests__/coin-tester/scenarios/ethereum.ts:247]: And same logic here, since we checked every transaction for its impact on account, we can just check something like:
Account should now have X ops, none of them have failed and balance is now: Y (TBD if we need to check more as previous expects already did the job I think)

File: libs/coin-modules/coin-evm/src/__tests__/coin-tester/scenarios/polygon.ts
Comment (lambertkevin) [inline libs/coin-modules/coin-evm/src/__tests__/coin-tester/scenarios/polygon.ts:1]: Same thing here in general

File: libs/coin-tester/docker-compose.yml [hunk 36]
```diff
@@ -0,0 +1,15 @@
+version: '3.8'
```
Comment (lambertkevin) [inline libs/coin-tester/docker-compose.yml:1]: Same thing about version to be checked

File: libs/coin-tester/docker-compose.yml [hunk 37]
```diff
@@ -0,0 +1,15 @@
...
+
+services:
+  speculos:
+    # Speculos isn't supported for Mac M1 yet. You have to build the image locally as described in the README.
+    # Otherwise you can use the official image
+    # image: ghcr.io/ledgerhq/speculos:latest
+    image: speculos:latest
```
Comment (lambertkevin) [inline libs/coin-tester/docker-compose.yml:8]: Do we have a way to condition this based on a value in env instead ? Because the CI won't be able use the container this way 🤔

File: libs/coin-tester/src/main.ts [hunk 38]
```diff
@@ -0,0 +1,208 @@
...
+    account: Account;
+    testTimeout?: number;
+    retryInterval?: number;
+    onSignerConfirmation?: (e?: SignOperationEvent) => Awaited<void>;
+  }>;
+  transactions: ScenarioTransaction<T>[];
+  beforeAll?: (account: Account) => Awaited<void>;
+  afterAll?: (account: Account) => Awaited<void>;
+  beforeEach?: (account: Account) => Awaited<void>;
+  afterEach?: (account: Account) => Awaited<void>;
```
Comment (lambertkevin) [inline libs/coin-tester/src/main.ts:30]: Can you explain why we need those `Awaited` utility type ?

**Review by lambertkevin — COMMENTED** [review]
Almost there, sorry for the new batch of requests, the only "major" ones are the expects not really testing what we want imo. We can discuss this if you want 👍
Everything else is just nice to have 👌

File: libs/coin-modules/coin-evm/src/__tests__/coin-tester/anvil.ts (hunk 28 above)
Comment (lvndry) [reply libs/coin-modules/coin-evm/src/__tests__/coin-tester/anvil.ts:1]: Good point. The library doesn't support v3 yet should we probably stick to `2.4` actually

File: libs/coin-modules/coin-evm/src/__tests__/coin-tester/.env.example [hunk 39]
```diff
@@ -0,0 +1,6 @@
+# https://iancoleman.io/bip39/
+SEED=your 12/15/24 words
+# has to be a token classic
+GH_TOKEN=gh_p
+API_PORT=4040
+SPECULOS_IMAGE=speculos
```
Comment (lambertkevin) [inline libs/coin-modules/coin-evm/src/__tests__/coin-tester/.env.example:6]: Maybe just add a comment here to say that this is only necessary for Apple Chip

File: libs/coin-modules/coin-evm/src/__tests__/coin-tester/scenarios.test.ts [hunk 40]
```diff
@@ -0,0 +1,42 @@
...
+import { killAnvil } from "./anvil";
+import { scenarioEthereum } from "./scenarios/ethereum";
+import { scenarioPolygon } from "./scenarios/polygon";
+
+global.console = require("console");
+jest.setTimeout(100_000);
+
+afterAll(async () => {
+  await Promise.all([killSpeculos(), killAnvil()]);
+});
```
Comment (lambertkevin) [inline libs/coin-modules/coin-evm/src/__tests__/coin-tester/scenarios.test.ts:12]: Is this expected to have this outside of the `describe` ?

File: libs/coin-modules/coin-evm/src/__tests__/coin-tester/scenarios/ethereum.ts [hunk 41]
```diff
@@ -0,0 +1,263 @@
...
+      expect(
+        latestOperation?.nftOperations?.find(
+          op => op.contract === "0xBC4CA0EdA7647A8aB7C2061c2E118A18a936f13D",
+        ),
+      ).toBeDefined();
+      expect(
+        latestOperation.nftOperations?.find(
+          operation => operation.contract === "0xBC4CA0EdA7647A8aB7C2061c2E118A18a936f13D",
+        ),
+      ).toBeDefined();
```
Comment (lambertkevin) [inline libs/coin-modules/coin-evm/src/__tests__/coin-tester/scenarios/ethereum.ts:86]: Loops are probably overkill, `const lastNftOperation = lastOperation?.nftOperations[0]` and you can expect it to equal an `expect.objectContaining({ contract: "0xBC4CA0EdA7647A8aB7C2061c2E118A18a936f13D", tokenId: "3368", value: new BigNumber(1) })`

File: libs/coin-modules/coin-evm/src/__tests__/coin-tester/scenarios/ethereum.ts [hunk 42]
```diff
@@ -0,0 +1,263 @@
...
+      expect(
+        latestOperation?.nftOperations?.find(
+          op => op.contract === "0x348FC118bcC65a92dC033A951aF153d14D945312",
+        ),
+      ).toBeDefined();
+      const latestNFTOperation = latestOperation.nftOperations?.find(
+        operation => operation.contract === "0x348FC118bcC65a92dC033A951aF153d14D945312",
+      );
+      expect(latestNFTOperation).toBeDefined();
+      expect(latestNFTOperation?.value.toFixed()).toBe("2");
```
Comment (lambertkevin) [inline libs/coin-modules/coin-evm/src/__tests__/coin-tester/scenarios/ethereum.ts:118]: Same thing here, we can make it simpler 👍

File: libs/coin-modules/coin-evm/src/__tests__/coin-tester/scenarios/ethereum.ts [hunk 43]
```diff
@@ -0,0 +1,263 @@
...
+    type: "0x0",
+    gasPrice: (await provider.getGasPrice()).toHexString(),
+    nonce: "0x" + (await provider.getTransactionCount(addressToImpersonnate)).toString(16),
+    chainId: "0x" + (await provider.getNetwork()).chainId.toString(16),
+  };
+
+  const hash = await provider.send("eth_sendTransaction", [impersonatedAccount]);
+  await provider.send("anvil_stopImpersonatingAccount", [addressToImpersonnate]);
+  await provider.waitForTransaction(hash);
+};
```
Comment (lambertkevin) [inline libs/coin-modules/coin-evm/src/__tests__/coin-tester/scenarios/ethereum.ts:157]: Nice 👏 👍

File: libs/coin-modules/coin-evm/src/__tests__/coin-tester/scenarios/ethereum.ts [hunk 44]
```diff
@@ -0,0 +1,263 @@
...
+    return {
+      currencyBridge,
+      accountBridge,
+      account: scenarioAccount,
+      onSignerConfirmation,
+    };
+  },
+  beforeAll: account => {
+    expect(account.nfts?.length).toBe(2);
+  },
```
Comment (lambertkevin) [inline libs/coin-modules/coin-evm/src/__tests__/coin-tester/scenarios/ethereum.ts:253]: Missing account.balance = 10k eth + 1 tokenAccount for USDC with balance = 100 usdc

File: libs/coin-modules/coin-evm/src/__tests__/coin-tester/scenarios/polygon.ts
Comment (lambertkevin) [inline libs/coin-modules/coin-evm/src/__tests__/coin-tester/scenarios/polygon.ts:1]: Same logic to apply here 🙏 
Also we never had ERC721/ERC1155 transactions here ? 🤔

**Review by lambertkevin — APPROVED** [review]
Great job ! 

Just minor things to change, and you're good to merge this 👍

**Review by lambertkevin — APPROVED** [review]
🚀
